### PR TITLE
ARROW-7824: [C++][Dataset] WriteFragments to disk

### PR DESCRIPTION
**Pull request by bkietz**
First draft of writing Fragments to disk. Writing a fragment to disk requires a FileFormat (in which the batches will be written) and a destination FileSource.

FileSystemSource::Write is provided as a convenience method: it preserves partition information while minimally changing file names.

### CODE REVIEW

**Comment by kszucs** [comment]
@bkietz requires a rebase

File: cpp/src/arrow/scalar.cc [hunk 1]
```diff
@@ -40,6 +41,90 @@ using internal::checked_pointer_cast;
...
+    return Status::OK();
+  }
+
+  Status ArrayHash(const Array& a) { return ArrayHash(*a.data()); }
+
+  Status ArrayHash(const ArrayData& a) {
+    RETURN_NOT_OK(StdHash(a.length) & StdHash(a.GetNullCount()));
+    for (const auto& buffer : a.buffers) {
+      if (buffer != nullptr) {
+        // FIXME(bkietz) this relies on values under nulls
```
Comment (fsaintjacques) [inline cpp/src/arrow/scalar.cc]: This will make valgrind/ubsan unhappy.

File: cpp/src/arrow/scalar_test.cc [hunk 2]
```diff
@@ -81,6 +82,20 @@ TYPED_TEST(TestNumericScalar, Basics) {
   ASSERT_EQ(*null_value, *dyn_null_value);
 }
 
+TYPED_TEST(TestNumericScalar, Hashing) {
+  using T = typename TypeParam::c_type;
+  using ScalarType = typename TypeTraits<TypeParam>::ScalarType;
+
+  std::unordered_set<std::shared_ptr<Scalar>, Scalar::Hash, Scalar::PtrsEqual> set;
```
Comment (fsaintjacques) [inline cpp/src/arrow/scalar_test.cc]: Good to know.

File: cpp/src/arrow/scalar.cc [hunk 3]
```diff
@@ -40,6 +41,90 @@ using internal::checked_pointer_cast;
...
+    for (const auto& child : a.child_data) {
+      RETURN_NOT_OK(ArrayHash(*child));
+    }
+    return Status::OK();
+  }
+
+  explicit ScalarHashImpl(const Scalar& scalar) { AccumulateHashFrom(scalar); }
+
+  void AccumulateHashFrom(const Scalar& scalar) {
+    DCHECK_OK(StdHash(scalar.type->fingerprint()));
```
Comment (fsaintjacques) [inline cpp/src/arrow/scalar.cc]: Just ensure that this doesn't take the metadata into account.

Comment (bkietz) [reply cpp/src/arrow/scalar.cc]: It doesn't; there's a separate `metadata_fingerprint()` method

File: cpp/src/arrow/dataset/writer.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  WriteTask(FileSource destination, std::shared_ptr<FileFormat> format)
+      : destination_(std::move(destination)), format_(std::move(format)) {}
+
+  Status CreateDestinationParentDir() const;
+
+  FileSource destination_;
+  std::shared_ptr<FileFormat> format_;
+};
+
+/// \brief A declarative plan for writing fragments to a partitioned directory structure.
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/writer.h]: will do

File: cpp/src/arrow/dataset/writer.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace arrow {
+namespace dataset {
+
+// XXX does this merit its own header? Maybe this should just be folded into file_base.h
+
+/// \brief Write a fragment to a single OutputStream.
+class ARROW_DS_EXPORT WriteTask {
+ public:
+  virtual Result<std::shared_ptr<FileFragment>> Execute() = 0;
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/writer.h]: okay

File: cpp/src/arrow/dataset/writer.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  std::shared_ptr<fs::FileSystem> filesystem;
+  std::string partition_base_dir;
+
+  using FragmentOrPartitionExpression =
+      util::variant<std::shared_ptr<Expression>, std::shared_ptr<Fragment>>;
+
+  /// If fragment_or_partition_expressions[i] is a Fragment, that Fragment will be
+  /// written to paths[i]. If it is an Expression, a directory representing that partition
+  /// expression will be created at paths[i] instead.
+  std::vector<FragmentOrPartitionExpression> fragment_or_partition_expressions;
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/writer.h]: The result of executing a `WritePlan` is a `FileSystemDataset`. My assumption is that when performing an extraction of a subset the extracted data is going to be used almost immediately as a dataset, so going through discovery again would be unnecessary. Thus all of the elements necessary to construct a Dataset around the written data are kept around.

FileSystemDataset must be constructed with directories as well as file names to correctly handle partition column materialization and fragment pruning. The directory `/base_dir/a=3` serves as a holder for the partition expression `"a"_ == 3`, which is applied to child directories and fragments.

File: cpp/src/arrow/scalar.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/scalar.cc]: Will fix

File: cpp/src/arrow/dataset/writer.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/writer.h]: (the same logic applies to WriteTask::Execute returning a Result<Fragment> rather than just Status)

**Comment by bkietz** [comment]
@fsaintjacques I have addressed your comments, PTAL

**Review by fsaintjacques — APPROVED** [review]

File: cpp/src/arrow/dataset/partition.cc [hunk 7]
```diff
@@ -192,10 +282,232 @@ class DirectoryPartitioningFactory : public PartitioningFactory {
...
+  Result<WritePlan> MakeWritePlan(FragmentIterator fragments) override;
+
+  Result<WritePlan> MakeWritePlan(FragmentIterator fragments,
+                                  const std::shared_ptr<Schema>& schema) override;
+
  private:
   std::vector<std::string> field_names_;
 };
 
+struct DirectoryPartitioningFactory::MakeWritePlanImpl {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/partition.cc]: Will do

[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/dataset/partition.cc]: @fsaintjacques PTAL

**Pull request closed without merging**